Newsletter categories: add subkey authentication for `useSubscribedNewsletterCategories` `useNewsletterCategorySubscriptionMutation`

TimBroddin opened the pull request:
## Proposed Changes

This PR adds subkey authentication for 2 endpoints that should be accessible from the Subscription Management portal.

Authentication in this portal is done by setting a subkey cookie. To be able to query API endpoints, this query needs to be passed in an Authorization header with `X-WPSUBKEY <subkey>` value.

There is already a similar method present in Reader, but it didn't feel best practice to import something from a totally unrelated part of Calypso. A follow-up project will be to refactor the queries that use it there to use this new TypeScript implementation.

## Testing Instructions

1. Apply this PR
2. Subscribe to a site with newsletter categories enabled, as a non-WPCOM user
3. Confirm the subscription
4. Do the subkey cookie dance:
- copy the value of the cookie, 
- Navigate to http://calypso.localhost:3000/subscriptions
- Paste `document.cookie = "subkey=VALUE"` into your console
- Refresh & behold your subscriptions
5. Click the site you've just subscribed to
6. You should see the categories listed:

![CleanShot 2023-09-21 at 16 26 18@2x](https://github.com/Automattic/wp-calypso/assets/528287/33833c7c-b660-4ac2-b64b-681c5276a8a1)

7. Change some categories, they should update

Now repeat for a WPCOM user, both in /subscriptions and Reader

---

By the way, you might see this in the console, this is not caused by this PR - but was already present:

![CleanShot 2023-09-21 at 16 22 14@2x](https://github.com/Automattic/wp-calypso/assets/528287/0d2b6aca-a491-4360-b0c5-32a1a4ff2bed)


## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

TimBroddin committed:
Added a check for user authentication before mutating or querying newsletter category subscriptions. Created two new helper functions: `useIsLoggedIn` to determine if the user is authenticated, and `requestWithSubkeyFallback` to handle requests with a fallback for users who are logged in with a valid subscription key. m.

TimBroddin committed:
Move to lib

TimBroddin committed:
Fix test

phcp reviewed (approved):
PR works as expected, now I can see the category toggles on the site subscription page. Let's just wait for the unit test fixes before merging. Good job @TimBroddin!

TimBroddin committed:
Fix test

TimBroddin committed:
Declare subscriptionManagementSubkey on window.currentUser

TimBroddin committed:
Add local subkey window definition to fix test fails

Replaced the global Window type definition with a local SubkeyWindow type in the request-with-subkey-fallback module. This was done to circumvent test failures caused by overriding the Window type in data-stores/src/user/resolvers.ts. Now the subscriptionManagementSubkey attribute is defined locally, ensuring no conflicts with other Window type definitions.
